src/bvals/cc/bvals_utils.hpp

```diff
@@ -96,8 +97,33 @@ inline void ForEachBoundary(std::shared_ptr<MeshData<Real>> &md, F func) {
             // No flux correction required unless boundaries share a face
             if (std::abs(nb.ni.ox1) + std::abs(nb.ni.ox2) + std::abs(nb.ni.ox3) != 1)
               continue;
+          } else if constexpr (bound == BoundaryType::restricted) {
+            // Check if restriction is required
+            if (nb.snb.level >= pmb->loc.level) continue;
+          }
+          if constexpr (bound == BoundaryType::restricted) {
+            IndexRange bni, bnj, bnk;
+            ComputeRestrictionBounds(bni, bnj, bnk, nb, pmb);
+            // This loop is only over {-1, 0, 1}^3 at most
+            for (int nk = bnk.s; nk <= bnk.e; ++nk) {
+              for (int nj = bnj.s; nj <= bnj.e; ++nj) {
+                for (int ni = bni.s; ni <= bni.e; ++ni) {
+                  int ntype = std::abs(ni) + std::abs(nj) + std::abs(nk);
+                  // skip myself or coarse levels; only the same level must be restricted
+                  if (ntype == 0 ||
+                      pmb->pbval->nblevel[nk + 1][nj + 1][ni + 1] != pmb->loc.level)
+                    continue;
+                  OffsetIndices offsets(nk, nj, ni);
+                  if (func_caller(func, pmb, rc, nb, v, offsets) ==
+                      LoopControl::break_out)
+                    return;
+                }
+              }
+            }
```

Comment:
> It appears that this can introduce up to nine visits to the function for a single boundary element, which means that the function is not really being called for each boundary but possibly multiple times for each boundary. 

It introduces up to eight visits... the 0,0,0 val is never called. But I *think* this is by design. The sub-pieces of each boundary halo are accessed. This is different than tying explicitly to buffers. That said, I have to admit, this was copied out of the original Athena++ code and I don't fully understand it.

> I would advocate for having this internal looping be done explicitly inside the kernel

I first wanted to put this loop in the kernel, but I can't actually because of the `continue`.

> or create a new ForEachBoundaryAndOffset method.

A new `ForEachBoundaryAndOffset` function would work, but it'll end up resulting in more repeated code. There would need to be a new `RebuildBufferCache` function for example. If that's the path we want to go down, I can be convinced, but we'd basically be copy-pasting two identical functions. I agree including the `offsets` in all the other functions is gross.

I guess in the longer term, rewriting the indexing logic to be less opaque might also resolve the issue, but I'd rather punt on that.

Reply:
> It introduces up to eight visits... the 0,0,0 val is never called. But I think this is by design. The sub-pieces of each boundary halo are accessed. This is different than tying explicitly to buffers. That said, I have to admit, this was copied out of the original Athena++ code and I don't fully understand it.

Confirmed this is by design. What's happening here is the restriction isn't over the same shape as the boundaries for ghost halo exchange. It's over whatever indices in the coarse buffer are required to provide enough information for prolongation, which may contain pieces of the interior and also contains things like edges and corners of the meshblock.

I don't know what that implies about what we should choose for the `ForEachBoundary` function. I guess we need to decide: Is `ForEachBoundary` only for comms, or is this infrastructure more general?

Reply:
After our discussion yesterday, I think we should leave this as is for this PR and think about simplifying things in a future PR. I am fairly convinced that this is more complicated than it needs to be (and that the ghost restriction step just needs to be done over some subset of the topological elements of a block). Approving the PR.

Reply:
:+1: Sounds good. I agree and I am looking at simplifying this in a future PR.